### ISSUE TITLE
Loosen dependency version constraints to permit flexible dependency resolution

### ISSUE THREAD

**Issue by chuckwondo** [MAAP-Project/maap-py#76]
For this project, dependencies are specified in `setup.py`, but there are 3 issues with how they are specified:

1. Most of them include _no_ version constraint. In general, all dependencies should at least be constrained to a major version to avoid unintended breakages from introduction of new major version releases
2. Only 2 include a version constraint, and they are [version matching](https://peps.python.org/pep-0440/#version-matching) clauses (`==`, i.e., _exact_ version match), which is rarely necessary, and can often be detrimental to dependency resolution. Instead, [compatible release](https://peps.python.org/pep-0440/#compatible-release) clauses (`~=`, flexible, but constrained to a particular _major_ version only) are generally preferred.
3. Several dev/test dependencies are listed in the _required_ dependencies, which is undesirable. They should be moved from `install_requires` to `test_requires`.

We should clean up these dependencies as outlined above, to allow for more flexible dependency resolution, while constraining to major versions to avoid potential breakages when new major dependency versions with breaking changes are released.

**Comment by chuckwondo** [comment]
@grallewellyn, I noticed that you've made recent changes to the dependencies in `setup.py`, so I wanted to make sure you're aware of this issue.

I'm going to restructure the dependencies, so let me know what specific problems you encountered previously with dependencies. I want to make sure I address the issue described here without causing any regressions.

**Comment by grallewellyn** [comment]
Hi Chuck, thank you for doing this!
1. I agree and we have discussed needing to pin versions, I can work on this 
2. I agree we should switch to `~=`, the strict version matching was implemented to fix the bug I explain below 
3. I agree 
The reason I pinned those versions is because we were getting dependencies errors trying to build the base images (vanilla, isce3, etc.) between the packages installed from our [environment.yml](https://github.com/MAAP-Project/maap-workspaces/blob/main/base_images/vanilla/environment.yml) and installing maap-py
The errors were:
```
awscli 2.14.1 requires cryptography<40.0.2,>=3.3.2, but you have cryptography 40.0.2 which is incompatible.
awscli 2.14.1 requires prompt-toolkit<3.0.39,>=3.0.24, but you have prompt-toolkit 3.0.43 which is incompatible.
aiobotocore 2.9.0 requires botocore<1.33.14,>=1.33.2, but you have botocore 1.34.3 which is incompatible.
```
I can work on this and change boto3 and ipython to be flexible versions `~=` and pin the rest of the versions 

Loosely related tangent, we plan to deploy maap-py as pip package soon

**Comment by chuckwondo** [comment]
@grallewellyn, thanks for the details.

However, I'm a bit puzzled by why this occurs:

```plain
aiobotocore 2.9.0 requires botocore<1.33.14,>=1.33.2, but you have botocore 1.34.3 which is incompatible.
```

Because in the `environment.yml` file that you link, I see [this line](https://github.com/MAAP-Project/maap-workspaces/blob/main/base_images/vanilla/environment.yml#L9):

```plain
  - boto3=1.34.15
```

How is it that the error message indicates 1.34.3 as the problem, and not 1.34.15? This is another file where you likely want to use flexible version pins, not exact version pins.

**Comment by grallewellyn** [comment]
@anilnatha What do you think about changing our environment.ymls to be using flexible version pins? 

Even though in our environment.yml (which installs before the maap-py install) we are installing boto3 1.34.15, when it went to install boto3 from the maap-py setup.py, it changed the version. I tried first to resolve these dependency issues by altering our environment.yml, but I couldn't get that to work so I had to change the setup.py. I was also puzzled by that error

**Comment by chuckwondo** [comment]
> the only potential downside is if the package versions change slightly between releases, and aren't backwards compatible, this could be annoying for users

If you pin to major versions, this is not a problem. For example, instead of `boto3=1.34.15`, you would loosen this to `boto3~=1.33.0`, which would pin only the _major_ version to `1`, but allow any minor release >= 33 to be used in order to resolve dependencies. There should be no breaking changes introduced by minor/patch versions. Of course, it's possible for that to happen by accident, but the likelihood is low, and if that happens, you can then add a constraint to specifically _exclude_ such releases (e.g., `boto3~=1.33.0,!=1.34.20`, if 1.34.20 included a breaking change by accident).

Also, why is `maap-py` installed separately? Why is it not included in `environment.yml`?

**Comment by grallewellyn** [comment]
We don't have plans to test all packages for each environment, so breaking changes would need to be reported by affected users. Other than that, this sounds good 

We were getting errors when we tried putting maap-py in the environment.yml. Keeping it outside the enivironment.yml is just temporary until we deploy maap-py as a pip package

**Comment by chuckwondo** [comment]
What errors were you getting with maap-py in environment.yml?

**Comment by grallewellyn** [comment]
The maap.cfg file necessary for maap-py wasn't being created so maap-py wasn't working correctly when we tried to add `- git+https://github.com/MAAP-Project/maap-py.git@v3.1.3#egg-info=maappy` to the environment.yml instead of the dockerfile

**Comment by chuckwondo** [comment]
I don't understand. How does the `maap.cfg` file come into play during dependency resolution/installation? Can you be more specific about exactly what problems occurred and when/where they occurred? Did you see an error during dependency installation when maappy was in environment.yml, or at some other point?

BTW, you should drop the `#egg-info=maappy` suffix. It is discouraged in this case. See https://pip.pypa.io/en/stable/topics/vcs-support/#url-fragments.

**Comment by grallewellyn** [comment]
When you instantiate MAAP in the ADE like 
```
from maap.maap import MAAP
maap = MAAP(maap_host='api.maap-project.org')
```
It needs to know where the maap.cfg is in this line https://github.com/MAAP-Project/maap-py/blob/master/maap/maap.py#L39
We have an [issue](https://github.com/MAAP-Project/Community/issues/909) to remove the need for locating the maap.cfg file 

The issue that was occurring was that when you try to instantiate MAAP() in a notebook in the ADE, you would get an error "No maap.cfg file found." The image build would be successful though

**Comment by chuckwondo** [comment]
I still don't follow. How does moving maap-py out of environment.yml and doing a separate maap-py install make any difference in that scenario? What process puts the maap.cfg file in place?

**Comment by grallewellyn** [comment]
The difference was probably that we couldn't find the location of the maap-py folder when we installed maap-py from the environment.yml and it didn't make sense to continue testing when we are planning to move to a different way for installation in the next couple releases. 
I believe the maap.cfg file is just installed where maap-py is installed, so when you set the environment variable, you just need to know where maap-py got installed then maap.cfg is inside maap-py

**Comment by chuckwondo** [comment]
Oooh! So the maap.cfg file that's in the repo is the actual maap.cfg file that is used? That file contains a token. That seems like an exposed sensitive value, no?

**Comment by grallewellyn** [comment]
Yes

@sujen1412 @marjo-luc Is maap_token exposed sensitive data?

**Comment by sujen1412** [comment]
That token is not being used for a while now. This would be fixed in favor of deprecating the cfg file altogether.

**Comment by chuckwondo** [comment]
Fixed by #77
